In [3]:
import pandas as pd
import numpy as np
dataframe = pd.read_csv("/content/drive/MyDrive/Worksheet5/student.csv")

In [4]:
#To-Do1
dataframe.head(5)


Math  Reading  Writing
0    48       68       63
1    62       81       72
2    79       80       78
3    76       83       79
4    59       64       62

In [5]:
print(dataframe.tail(5))

     Math  Reading  Writing
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72


In [6]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   Math     1000 non-null   int64
 1   Reading  1000 non-null   int64
 2   Writing  1000 non-null   int64
dtypes: int64(3)
memory usage: 23.6 KB


In [7]:
dataframe.describe()

Math      Reading      Writing
count  1000.000000  1000.000000  1000.000000
mean     67.290000    69.872000    68.616000
std      15.085008    14.657027    15.241287
min      13.000000    19.000000    14.000000
25%      58.000000    60.750000    58.000000
50%      68.000000    70.000000    69.500000
75%      78.000000    81.000000    79.000000
max     100.000000   100.000000   100.000000

In [8]:
X = dataframe.iloc[:, :-1]

Y = dataframe.iloc[:, -1]

print("Features (X):")
print(X.head())

print("\nLabel (Y):")
print(Y.head())


Features (X):
   Math  Reading
0    48       68
1    62       81
2    79       80
3    76       83
4    59       64

Label (Y):
0    63
1    72
2    78
3    79
4    62
Name: Writing, dtype: int64


In [9]:
#To-Do2
print(f"X shape: {X.shape}")
print(f"Y shape: {Y.shape}")


X shape: (1000, 2)
Y shape: (1000,)


In [10]:
#To-Do3
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, Y, test_size=0.3, random_state=42
)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)


X_train shape: (700, 2)
X_test shape: (300, 2)
y_train shape: (700,)
y_test shape: (300,)


In [11]:
#To-Do4
def cost_function(X, Y, W):
  X = np.array(X, dtype=float)
  Y = np.array(Y, dtype=float).reshape(-1, 1)
  W = np.array(W, dtype=float).reshape(-1, 1)
  n= len(Y)
  Yi= np.dot(X,W) #Can also be used as X @ W
  cost= (1/2*n)*np.sum((Yi-Y)**2)
  return cost

In [12]:
#To-Do5
X_test = np.array([[1, 2], [3, 4], [5, 6]])
Y_test = np.array([3, 7, 11])
W_test = np.array([1, 1])
cost = cost_function(X_test, Y_test, W_test)
if cost == 0:
  print("Proceed Further")
else:
  print("something went wrong: Reimplement a cost function")
print("Cost function output:", cost_function(X_test, Y_test, W_test))

Proceed Further
Cost function output: 0.0


In [13]:
#To-Do6
def gradient_descent(X, Y, W, alpha, iterations):
  cost_history = [0] * iterations
  m = len(Y)
  W_update = W.copy()
  for iteration in range(iterations):
    Y_pred = X.dot(W_update)
    loss = Y_pred - Y
    dw = (1/m)*X.T.dot(loss)
    W_update = W_update - alpha * dw
    cost = cost_function(X, Y, W_update)
    cost_history[iteration] = cost
  return W_update, cost_history


In [14]:
#To-Do7
np.random.seed(0)
X = np.random.rand(100, 3)
Y = np.random.rand(100)
W = np.random.rand(3)

alpha = 0.01
iterations = 1000

final_params, cost_history = gradient_descent(X, Y, W, alpha, iterations)

print("Final Parameters:", final_params)
print("Cost History:", cost_history)

Final Parameters: [0.20551667 0.54295081 0.10388027]
Cost History: [np.float64(1071.1197094660154), np.float64(1063.4880599939902), np.float64(1055.9826315680616), np.float64(1048.6012948320558), np.float64(1041.3419564285339), np.float64(1034.2025583900624), np.float64(1027.181077540776), np.float64(1020.2755249080619), np.float64(1013.4839451441931), np.float64(1006.8044159577371), np.float64(1000.2350475545869), np.float64(993.7739820884376), np.float64(987.4193931205608), np.float64(981.1694850887098), np.float64(975.0224927850095), np.float64(968.976680842672), np.float64(963.0303432313951), np.float64(957.1818027612914), np.float64(951.4294105952065), np.float64(945.7715457692842), np.float64(940.2066147216397), np.float64(934.7330508290016), np.float64(929.3493139511913), np.float64(924.0538899833018), np.float64(918.8452904154543), np.float64(913.722051899995), np.float64(908.6827358260123), np.float64(903.7259279010502), np.float64(898.8502377398918), np.float64(894.0542984603

In [15]:
#To-Do8
def rmse(Y, Y_pred):
  rmse=np.sqrt(np.mean((Y-Y_pred)**2))
  return rmse

In [16]:
#To-Do9
def r2(Y, Y_pred):
  mean_y=np.mean(Y)
  ss_tot=np.sum((Y - mean_y)**2)
  ss_res=np.sum((Y-Y_pred)**2)
  r2=1-(ss_res/ss_tot)
  return r2

In [22]:
#To-Do10
def main():
  data = pd.read_csv("/content/drive/MyDrive/Worksheet5/student.csv")
  X = data[['Math', 'Reading']].values
  Y = data['Writing'].values
  X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=42)
  W = np.zeros(X_train.shape[1])
  alpha = 0.00001
  iterations = 1000
  W_optimal, cost_history = gradient_descent(X_train, Y_train, W, alpha, iterations)
  Y_pred= np.dot(X_test, W_optimal)
  model_rmse = rmse(Y_test, Y_pred)
  model_r2 = r2(Y_test, Y_pred)
  print("Final Weights: ", W_optimal)
  print("Cost History (First 10 iterations):" , cost_history[:10])
  print("RMSE on test set: ", model_rmse)
  print("R-Squared on test set: ", model_r2)
if __name__ == "__main__":
  main()


Final Weights:  [0.34973248 0.64484523]
Cost History (First 10 iterations): [np.float64(985837985.1820028), np.float64(803459926.1953541), np.float64(655109165.8633409), np.float64(534436943.30557907), np.float64(436279039.7322294), np.float64(356434768.39319396), np.float64(291487200.24442816), np.float64(238656931.63436297), np.float64(195683135.7916454), np.float64(160726807.54429546)]
RMSE on test set:  5.224589785890595
R-Squared on test set:  0.8868710151151797


To-Do11
1. An R-squared value of 0.88 shows that the model predicts the output very well, whereas an RMSE of 5.225 indicates that the prediction errors are small. As a result, neither underfitting nor overfitting affect the model's performance.

2. According to experiments using various learning rates, large values caused fast divergence, while very small values caused slow convergence, . The optimal balance was obtained with a learning rate of 0.00001, which led to faster convergence, a lower RMSE, and a higher R-squared value, all of which enhanced model performance.